### PR TITLE
[#12048] v9: Skeleton implementation

### PR DESCRIPTION
**Pull request by samuelfangjw**
Part of #12048

Libraries/Frameworks added:
- Hibernate
- Liquibase for sql migrations
- Mockito for unit tests
- Testcontainers for integration tests
- Related dependencies

Major Changes from POC:
1. Reworked logic classes to have dependencies injected (for future unit tests).
2. Removed *Attributes class.
4. Hibernate SessionFactory now set up in a ServletContextListener rather than a Filter. 
5. Hibernate begin session/transaction moved to servlet root (from filter). This way exceptions can be caught and rollback initiated.
6. Differentiate between Integration and Unit tests (currently following popular *Test.java and *IT.java convention).
7. Add gradle tasks to run unit and integration tests separately.

Liquibase:
To access liquidate commands, `./gradlew <command>`. A list of commands can be found by checking `./gradlew tasks`. In particular, `./gradlew liquibaseUpdate` to perform all migrations.

Changelogs are organised by release, with `db.changelog-v9` containing changelogs for v9, Unfortunately, the very useful liquidate-hibernate6 extension seems to be java 17 and above only. I have been generating changelogs by letting hibernate update the db then using `liquibaseGenerateChangelog`. Will find another flow to generate changelogs.

Sample implementation:
Partial implementation of course related api/logic/db.

### CODE REVIEW

**Review by zhaojj2209 — APPROVED** [review]
Set up and tested skeleton using postgres on local machine, will need someone else to help test using the docker image. Am able to create a course and have it persist in the DB.

I assume we'll leave fixing of component & E2E tests to future PRs?

**Comment by samuelfangjw** [comment]
> Set up and tested skeleton using postgres on local machine, will need someone else to help test using the docker image. Am able to create a course and have it persist in the DB.
> 
> I assume we'll leave fixing of component & E2E tests to future PRs?

I'll take a quick look at the tests tomorrow and fix those that I can. For the E2E test, I can fix the no transaction in session issue but cannot fix the test until GetCourseAction is implemented properly.

File: src/test/java/teammates/test/BaseTestCaseWithSqlDatabaseAccess.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        HibernateUtil.getSessionFactory().getCurrentSession().getTransaction().begin();
+        DbMigrationUtil.resetDb(PGSQL.getJdbcUrl(), PGSQL.getUsername(), PGSQL.getPassword());
+    }
+
+    @AfterMethod
+    public void tearDown() {
+        HibernateUtil.getSessionFactory().getCurrentSession().getTransaction().commit();
+    }
+
+    public void verifyEquals(BaseEntity expected, BaseEntity actual) {
```
Comment (hhdqirui) [inline src/test/java/teammates/test/BaseTestCaseWithSqlDatabaseAccess.java:54]: Changing to `protected` can prevent TestNG from recognizing this method as a test.
```suggestion
    protected void verifyEquals(BaseEntity expected, BaseEntity actual) {
```

**Review by hhdqirui — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/test/BaseTestCaseWithSqlDatabaseAccess.java (hunk 1 above)
Comment (samuelfangjw) [reply src/test/java/teammates/test/BaseTestCaseWithSqlDatabaseAccess.java:54]: Thank you! Was wondering why that was being recognised as a test.

**Review by hhdqirui — APPROVED** [review]
LGTM

**Pull request merged**